### PR TITLE
FP32 output layer matmul to compute vocab logits when residual path is in fp32

### PR DESCRIPTION
**Pull request by mkhona-nvidia**
# What does this PR do ?

The current code that keeps residual activations in fp32, does not do the output_layer GEMM in fp32 and instead downcasts to bf16. So we lose the mantissa bits of our fp32 activations residuals that we worked hard to maintain in the current implementation.

This adds the fp32 `output_layer` logits computation to mamba, gpt and mtp layers

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/utils.py [hunk 1]
```diff
@@ -112,6 +112,62 @@ def inner(func):
...
+
+    prev_val = torch.get_float32_matmul_precision()
+    torch.set_float32_matmul_precision(precision)
+    try:
+        yield
+    finally:
+        torch.set_float32_matmul_precision(prev_val)
+
+
+def compute_output_logits_fp32(
```
Comment (skyw) [inline megatron/core/utils.py:140]: This is too expensive in terms of both memory consumption and computation.

Better to wrap a gemm_ex directly with bf16 input and write out fp32, if pytorch doesn't provide anything like this.

File: megatron/core/models/gpt/gpt_model.py [hunk 2]
```diff
@@ -650,9 +651,17 @@ def _postprocess(
                     hidden_states.squeeze(1).unsqueeze(0)
                 ).unsqueeze(1)
 
-        logits, _ = self.output_layer(
-            hidden_states, weight=output_weight, runtime_gather_output=runtime_gather_output
-        )
+        if self.config.fp32_residual_connection:
```
Comment (rybakov) [inline megatron/core/models/gpt/gpt_model.py:654]: why flag is called fp32_residual_connection?
fp32_residual_connection is used only for setting output logit to fp32, but all residual are still in bf16?

Comment (rybakov) [reply megatron/core/models/gpt/gpt_model.py:654]: never mind, now I see that this MR is a follow up on top of https://github.com/NVIDIA/Megatron-LM/pull/3504/changes

Comment (mkhona-nvidia) [reply megatron/core/models/gpt/gpt_model.py:654]: I am in the process of rewriting this PR from scratch; the original intended use was only for keeping the `output_layer` in FP32 IF the residual was in FP32, but upon further thinking, I think there should be a separate flag to keep output logits in FP32 regardless of residual
